File: src/main/java/org/alpacaindustries/iremiaminigame/system/ui/GameScoreboard.java
package org.alpacaindustries.iremiaminigame.system.ui;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Criteria;
import org.bukkit.scoreboard.DisplaySlot;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Manages game-specific scoreboards with automatic player state preservation
 * Handles line management, team-based text display, and cleanup on player
 * disconnect
 */
public class GameScoreboard {
  private final Scoreboard scoreboard;
  private final Objective objective;
  private final Map<Integer, String> lines = new HashMap<>();
  private final Map<UUID, Scoreboard> previousScoreboards = new HashMap<>(); // Restore player's original scoreboard

  public GameScoreboard(Component title) {
    scoreboard = Bukkit.getScoreboardManager().getNewScoreboard();
    objective = scoreboard.registerNewObjective("game", Criteria.DUMMY, title);
    objective.setDisplaySlot(DisplaySlot.SIDEBAR);
  }

  /**
   * Set a line on the scoreboard using Adventure Component
   *
   * @param line      The line number (1-15, 1 being at the bottom)
   * @param component The component to display on this line
   * @return This scoreboard for chaining
   */
  /**
   * Array of color codes that can be used as invisible entry IDs
   * Using color codes makes them not visible in the scoreboard
   */
  private static final String[] COLOR_CODES = {
      "§0", "§1", "§2", "§3", "§4", "§5", "§6", "§7", "§8", "§9",
      "§a", "§b", "§c", "§d", "§e", "§f", "§r"
  };

  public GameScoreboard setLine(int line, Component component) {
    if (line < 1 || line > 15) {
      throw new IllegalArgumentException("Line must be between 1 and 15");
    }

    // Remove existing line if it exists
    if (lines.containsKey(line)) {
      scoreboard.resetScores(lines.get(line));
    }

    // Create an entry ID using the line number
    String entryId = "§" + Integer.toHexString(line); // Using hex color code as an invisible prefix

    // For lines > 15, we need to add extra characters to make unique ids
    if (line > 15) {
      entryId += "§r"; // Reset code as a spacer
    }

    lines.put(line, entryId);

    // Get or create the team for this line
    String teamName = "line_" + line;
    Team team = scoreboard.getTeam(teamName);
    if (team == null) {
      team = scoreboard.registerNewTeam(teamName);
    } else {
      // Clear any existing entries
      for (String entry : team.getEntries()) {
        team.removeEntry(entry);
      }
    }

    // Add our entry to the team
    team.addEntry(entryId);

    // Set the component text
    team.prefix(component);

    // Optionally add empty suffix to ensure consistent appearance
    team.suffix(Component.empty());

    // Set the score in the scoreboard
    objective.getScore(entryId).setScore(line);

    return this;
  }

  /**
   * Update a specific team's prefix/suffix
   * This is useful for updating values without flickering the scoreboard
   *
   * @param team   Team name
   * @param prefix New prefix for the team as a Component
   * @param suffix New suffix for the team as a Component
   * @return This scoreboard for chaining
   */
  public GameScoreboard updateTeam(String team, Component prefix, Component suffix) {
    Team scoreboardTeam = scoreboard.getTeam(team);
    if (scoreboardTeam == null) {
      scoreboardTeam = scoreboard.registerNewTeam(team);
      scoreboardTeam.addEntry(team);
    }
    scoreboardTeam.prefix(prefix);
    scoreboardTeam.suffix(suffix);

    return this;
  }

  /**
   * Update a specific team's prefix/suffix using legacy text format
   * This is useful for updating values without flickering the scoreboard
   *
   * @param team   Team name
   * @param prefix New prefix for the team (can include color codes with &)
   * @param suffix New suffix for the team (can include color codes with &)
   * @return This scoreboard for chaining
   */
  public GameScoreboard updateTeamLegacy(String team, String prefix, String suffix) {
    return updateTeam(
        team,
        LegacyComponentSerializer.legacyAmpersand().deserialize(prefix),
        LegacyComponentSerializer.legacyAmpersand().deserialize(suffix));
  }

  /**
   * Get the configured score for a team entry
   * Override this in subclasses if you want to customize the order
   *
   * @param team Team name
   * @return Score value for this team
   */
  protected int getTeamScore(String team) {
    // Default implementation uses the hashCode to get a "random" but consistent
    // value
    return Math.abs(team.hashCode()) % 15 + 1;
  }

  /**
   * Show this scoreboard to a player
   *
   * @param player Player to show the scoreboard to
   */
  public void showTo(Player player) {
    // Store their previous scoreboard
    previousScoreboards.put(player.getUniqueId(), player.getScoreboard());

    // Set our scoreboard
    player.setScoreboard(scoreboard);
  }

  /**
   * Hide this scoreboard from a player, restoring their previous one
   *
   * @param player Player to hide the scoreboard from
   */
  public void hideFrom(Player player) {
    // Restore their previous scoreboard if we have it
    Scoreboard previous = previousScoreboards.get(player.getUniqueId());
    if (previous != null) {
      player.setScoreboard(previous);
    } else {
      // Otherwise set them to the main scoreboard
      player.setScoreboard(Bukkit.getScoreboardManager().getMainScoreboard());
    }

    previousScoreboards.remove(player.getUniqueId());
  }

  /**
   * Get the underlying Bukkit scoreboard
   *
   * @return The Bukkit scoreboard
   */
  public Scoreboard getScoreboard() {
    return scoreboard;
  }
}

File: src/main/kotlin/org/alpacaindustries/iremiaminigamecore/system/ui/GameScoreboard.kt
package org.alpacaindustries.iremiaminigamecore.system.ui;

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.scoreboard.*
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * Manages game-specific scoreboards with automatic player state preservation.
 *
 * Features:
 * - Thread-safe operations
 * - Automatic cleanup on player disconnect
 * - Line-based and team-based content management
 * - Adventure Component support with legacy fallback
 * - Fluent API for easy chaining
 *
 * @param title The title component to display at the top of the scoreboard
 */
class GameScoreboard(title: Component) {

    private val scoreboard: Scoreboard = Bukkit.getScoreboardManager()!!.newScoreboard
    private val objective: Objective = scoreboard.registerNewObjective("game", Criteria.DUMMY, title)

    // Thread-safe collections for concurrent access
    private val lines = ConcurrentHashMap<Int, String>()
    private val previousScoreboards = ConcurrentHashMap<UUID, Scoreboard>()
    private val playerTeams = ConcurrentHashMap<UUID, Set<String>>()

    init {
        objective.displaySlot = DisplaySlot.SIDEBAR
    }

    companion object {
        private const val MAX_LINES = 15
        private const val MIN_LINES = 1

        /**
         * Creates a GameScoreboard with a legacy text title
         * @param title The title using legacy color codes (&)
         */
        fun withLegacyTitle(title: String): GameScoreboard {
            return GameScoreboard(LegacyComponentSerializer.legacyAmpersand().deserialize(title))
        }

        /**
         * Creates a GameScoreboard with a plain text title
         * @param title The plain text title
         */
        fun withPlainTitle(title: String): GameScoreboard {
            return GameScoreboard(Component.text(title))
        }
    }

    /**
     * Sets a line on the scoreboard using Adventure Component.
     *
     * @param line The line number (1-15, where 1 is at the bottom)
     * @param component The component to display on this line
     * @return This scoreboard for method chaining
     * @throws IllegalArgumentException if line is not between 1 and 15
     */
    fun setLine(line: Int, component: Component): GameScoreboard {
        require(line in MIN_LINES..MAX_LINES) {
            "Line must be between $MIN_LINES and $MAX_LINES, got $line"
        }

        // Remove existing line if it exists
        lines[line]?.let { entryId ->
            scoreboard.resetScores(entryId)
        }

        // Create a unique entry ID for this line
        val entryId = generateEntryId(line)
        lines[line] = entryId

        // Get or create the team for this line
        val teamName = "line_$line"
        val team = getOrCreateTeam(teamName)

        // Clear existing entries and add our new one
        team.entries.toList().forEach { team.removeEntry(it) }
        team.addEntry(entryId)

        // Set the component text as prefix
        team.prefix(component)
        team.suffix(Component.empty())

        // Set the score (higher scores appear at the top)
        objective.getScore(entryId).score = line

        return this
    }

    /**
     * Sets a line using legacy text format.
     *
     * @param line The line number (1-15)
     * @param text The text with legacy color codes (&)
     * @return This scoreboard for method chaining
     */
    fun setLineLegacy(line: Int, text: String): GameScoreboard {
        return setLine(line, LegacyComponentSerializer.legacyAmpersand().deserialize(text))
    }

    /**
     * Sets a line using plain text.
     *
     * @param line The line number (1-15)
     * @param text The plain text
     * @param color Optional color for the text
     * @return This scoreboard for method chaining
     */
    fun setLinePlain(line: Int, text: String, color: NamedTextColor? = null): GameScoreboard {
        val component = if (color != null) {
            Component.text(text, color)
        } else {
            Component.text(text)
        }
        return setLine(line, component)
    }

    /**
     * Removes a line from the scoreboard.
     *
     * @param line The line number to remove
     * @return This scoreboard for method chaining
     */
    fun removeLine(line: Int): GameScoreboard {
        lines[line]?.let { entryId ->
            scoreboard.resetScores(entryId)
            val teamName = "line_$line"
            scoreboard.getTeam(teamName)?.unregister()
            lines.remove(line)
        }
        return this
    }

    /**
     * Clears all lines from the scoreboard.
     *
     * @return This scoreboard for method chaining
     */
    fun clearLines(): GameScoreboard {
        lines.keys.toList().forEach { removeLine(it) }
        return this
    }

    /**
     * Updates a specific team's prefix and suffix.
     * This is useful for updating dynamic content without flickering.
     *
     * @param teamName Team name
     * @param prefix New prefix component
     * @param suffix New suffix component (optional)
     * @return This scoreboard for method chaining
     */
    fun updateTeam(
        teamName: String,
        prefix: Component,
        suffix: Component = Component.empty()
    ): GameScoreboard {
        val team = getOrCreateTeam(teamName)

        // Ensure the team has an entry if it doesn't already
        if (team.entries.isEmpty()) {
            team.addEntry(teamName)
        }

        team.prefix(prefix)
        team.suffix(suffix)

        return this
    }

    /**
     * Updates a team using legacy text format.
     *
     * @param teamName Team name
     * @param prefix New prefix with legacy color codes
     * @param suffix New suffix with legacy color codes (optional)
     * @return This scoreboard for method chaining
     */
    fun updateTeamLegacy(
        teamName: String,
        prefix: String,
        suffix: String = ""
    ): GameScoreboard {
        return updateTeam(
            teamName,
            LegacyComponentSerializer.legacyAmpersand().deserialize(prefix),
            LegacyComponentSerializer.legacyAmpersand().deserialize(suffix)
        )
    }

    /**
     * Creates a team for player-specific content.
     *
     * @param player The player
     * @param teamName Custom team name (optional, defaults to player UUID)
     * @param prefix Team prefix
     * @param suffix Team suffix (optional)
     * @return This scoreboard for method chaining
     */
    fun createPlayerTeam(
        player: Player,
        teamName: String = "player_${player.uniqueId}",
        prefix: Component = Component.empty(),
        suffix: Component = Component.empty()
    ): GameScoreboard {
        val team = getOrCreateTeam(teamName)
        team.addEntry(player.name)
        team.prefix(prefix)
        team.suffix(suffix)

        // Track teams for this player for cleanup
        playerTeams.compute(player.uniqueId) { _, existing ->
            (existing ?: emptySet()) + teamName
        }

        return this
    }

    /**
     * Shows this scoreboard to a player, preserving their previous scoreboard.
     *
     * @param player Player to show the scoreboard to
     */
    fun showTo(player: Player) {
        // Store their previous scoreboard for restoration
        previousScoreboards[player.uniqueId] = player.scoreboard
        player.scoreboard = scoreboard
    }

    /**
     * Shows this scoreboard to multiple players.
     *
     * @param players Collection of players
     */
    fun showTo(players: Collection<Player>) {
        players.forEach { showTo(it) }
    }

    /**
     * Hides this scoreboard from a player, restoring their previous one.
     *
     * @param player Player to hide the scoreboard from
     */
    fun hideFrom(player: Player) {
        val previous = previousScoreboards.remove(player.uniqueId)
        player.scoreboard = previous ?: Bukkit.getScoreboardManager()!!.mainScoreboard

        // Clean up player-specific teams
        cleanupPlayerTeams(player)
    }

    /**
     * Hides this scoreboard from multiple players.
     *
     * @param players Collection of players
     */
    fun hideFrom(players: Collection<Player>) {
        players.forEach { hideFrom(it) }
    }

    /**
     * Checks if a player is currently viewing this scoreboard.
     *
     * @param player The player to check
     * @return true if the player is viewing this scoreboard
     */
    fun isShownTo(player: Player): Boolean {
        return player.scoreboard == scoreboard
    }

    /**
     * Gets all players currently viewing this scoreboard.
     *
     * @return Set of players viewing this scoreboard
     */
    fun getViewers(): Set<Player> {
        return Bukkit.getOnlinePlayers().filter { it.scoreboard == scoreboard }.toSet()
    }

    /**
     * Updates the scoreboard title.
     *
     * @param title New title component
     * @return This scoreboard for method chaining
     */
    fun updateTitle(title: Component): GameScoreboard {
        objective.displayName(title)
        return this
    }

    /**
     * Updates the scoreboard title using legacy text.
     *
     * @param title New title with legacy color codes
     * @return This scoreboard for method chaining
     */
    fun updateTitleLegacy(title: String): GameScoreboard {
        return updateTitle(LegacyComponentSerializer.legacyAmpersand().deserialize(title))
    }

    /**
     * Cleans up all resources and hides the scoreboard from all viewers.
     */
    fun cleanup() {
        // Hide from all current viewers
        getViewers().forEach { hideFrom(it) }

        // Clear all data structures
        lines.clear()
        previousScoreboards.clear()
        playerTeams.clear()

        // Unregister the objective
        objective.unregister()
    }

    /**
     * Gets the underlying Bukkit scoreboard.
     * Use with caution as direct modifications may break the wrapper's functionality.
     *
     * @return The underlying Bukkit scoreboard
     */
    val bukkitScoreboard: Scoreboard
        get() = scoreboard

    // Private helper methods

    private fun generateEntryId(line: Int): String {
        // Use a combination of color codes to create unique, invisible entries
        val hexChar = Integer.toHexString(line).lowercase()
        return "§$hexChar§r"
    }

    private fun getOrCreateTeam(teamName: String): Team {
        return scoreboard.getTeam(teamName) ?: scoreboard.registerNewTeam(teamName)
    }

    private fun cleanupPlayerTeams(player: Player) {
        playerTeams.remove(player.uniqueId)?.forEach { teamName ->
            scoreboard.getTeam(teamName)?.let { team ->
                team.removeEntry(player.name)
                // Remove team if it has no more entries
                if (team.entries.isEmpty()) {
                    team.unregister()
                }
            }
        }
    }
}
